[MD] Tweak multiple data source design doc (#2724)
opensearch-project/OpenSearch-Dashboards: docs/multi-datasource/client_management_design.md

### 4.0 Answer some critical design questions

**1.** **How to set up connection(clients) for different data sources?**

> Similar to how current OpenSearch Dashboards talks to default OS by creating opensearch node.js client using [opensearch-js](https://github.com/opensearch-project/opensearch-js) library, for data sources we also create clients for each. Critical params that differentiate data sources are `url` and `auth`

Reviewer: If we are splitting data sources we might as well get the other things we missed like `OS` should be OpenSearch. 

> by creating opensearch node.js client using opensearch-js library

seems a little clunky. can just say by using creating a client using opensearch-js library.

> for data sources we also create clients for each *connection*

Reviewer: also is there any memory leakage with this design?

Author: Addressed. Gonna open another thread to discuss the memory leakage question.

  }
  ```

> - Get root client: Look up client Pool by **endpoint**, return client if existed. If it misses, we create new client instance and load into pool. At this step, the client won't have any auth info.

Reviewer 2: ```suggestion
- Get root client: Look up the client pool by **endpoint** and return the client if it exists. If a client was not found, a new client instance is created and loaded into pool. At this step, the client won't have any auth info.
```

- Get credentials: Call crypto service utilities to **decrypt** user credentials from `DataSource` Object.
- Assemble the actual query client: With auth info and root client, we’ll leverage the `opensearch-js` connection pooling strategy to create the actual query client from root client by `client.child()`.

#### 4.2.1 Legacy Client

> OpenSearch Dashboards is forked from Kibana 7.10. At the time of the fork happened, there are 2 types of client used in the codebase. One is the new client, which later was migrated as `opensearch-js`, the other one is the legacy client which is `elasticsearch-js`. Legacy clients are still used many critical features, such as visualization, index pattern management, along with new client.

Reviewer 2: ```suggestion
OpenSearch Dashboards was forked from Kibana 7.10 and at the time, the codebase made two types of clients available for use. One was the "new client" which has since been separated into `opensearch-js`, and the other was the legacy client named `elasticsearch-js`. Legacy clients are still used by some core features like visualization and index pattern management.
```

> OpenSearch Dashboards is forked from Kibana 7.10. At the time of the fork happened, there are 2 types of client used in the codebase. One is the new client, which later was migrated as `opensearch-js`, the other one is the legacy client which is `elasticsearch-js`. Legacy clients are still used many critical features, such as visualization, index pattern management, along with new client.

Reviewer: Tbh I think we should remove the mention of us forking from the legacy application.

Author: Removed

### 4.4 Refactor data plugin search module to call core API to get datasource client

> `Search strategy` is the low level API of data plugin search module. It retrieves clients and query OpenSearch. It needs to be refactored to switch between default client and datasource client, depending on whether a request is send to datasource or not.

Reviewer 2: ```suggestion
`Search strategy` is the low level API of data plugin search module. It retrieves clients and queries OpenSearch. It needs to be refactored to switch between the default client and the datasource client, depending on whether or not a request is sent to the datasource.
```

Reviewer: I think we also split "data source" in above paragrahp

client.search(params);
```

> Similarly we’ll have the following for datasource use case. `AsCurrentUser` is something does not make sense for datasource, because it’s always the “current” user credential defined in the “datasource”, that we are using to create the client, or look up the client pool.

Reviewer 2: ```suggestion
Similarly we’ll have the following for datasource use case. `AsCurrentUser` doesn't really apply to a datasource because it’s always the “current” user's credentials, defined in the “datasource”, that gets used to initialize the client or lookup the client pool.
```
